### ISSUE TITLE
Why check whether self.global_step is 0 in lorapro_full_adjustment method?

### ISSUE THREAD

**Issue by Frost-Mactavish** [mrflogs/LoRA-Pro#13]
I find the way of computing inverse matrix and calibrated gradient is different depending on whether `self.global_step` equals to 0 in `lorapro_full_adjustment` method. But `self.global_step` is initialized to 0 and will be incremented by 1 before `lorapro_full_adjustment` method is called in `step` method, meaning that `self.global_step` will never be 0.

```
def __init__(self, ...):
    ...
    self.global_step = 0

@torch.no_grad()
def lorapro_full_adjustment(self, partition_id):
    ...
    if self.global_step == 0:
        AA_T_inv = torch.linalg.pinv(AA_T + delta * torch.eye(A.shape[0], device=grad_A_orin.device)) 
        AA_T_inv = AA_T_inv.to(A.dtype)
        grad_A = grad_A_orin
        grad_B = (1 / scaling_factor ** 2) * grad_B_orin @ AA_T_inv
    else:
        AA_T_inv = torch.linalg.pinv(AA_T + delta * torch.eye(A.shape[0], device=grad_A_orin.device)) 
        B_TB_inv = torch.linalg.pinv(B_TB + delta * torch.eye(A.shape[0], device=grad_A_orin.device)) 
        ...

def step(self, closure=None):
    ...
    for i, group in enumerate(self.bit16_groups):
        self.timers(OPTIMIZER_GRADIENTS_TIMER).start()
        self.global_step += 1
        partition_id = dist.get_rank(group=self.real_dp_process_group[i])
        
        self.lorapro_full_adjustment(partition_id)
        ...
```

**So is the if statement necessary?** Since I know little about DeepSeed, does the framework have anything to do with it? Hope to get your kind explanation!

**Comment by LZY-the-boys** [comment]
Because B is zero when step is zero. B^-1 is illegal operation.

**Comment by mrflogs** [comment]
Thank you for pointing this out. I will fix this later.

When at step 0, we initialize the matrices as A=normal and B = 0. 

Under this initialization, the optimization problem becomes:
$$\min_{g_A, g_B} \\|\tilde{g} - g\\|^2_F, \quad \text{subject to} \quad \tilde{g} = s B g_A + s g_B A = s g_B A.$$
The closed-form solution for these conditions is:
$g^A = g^A_{lora}, g^B=\frac{1}{s^2}g^B(AA^T)^{-1}$.
In this case, there is no need to compute (B^TB)^{-1}.
